controller/volume_controller.go

```diff
@@ -267,16 +267,20 @@ func (vc *VolumeController) syncVolume(key string) (err error) {
 				}
 			}
 		}
-		// now engines have been deleted or in the process
-
 		for _, r := range replicas {
-			if r.DeletionTimestamp == nil {
-				if err := vc.ds.DeleteReplica(r.Name); err != nil {
-					return err
-				}
+			if err := vc.ds.DeleteReplica(r); err != nil {
+				return err
 			}
 		}
-		// now replicas have been deleted or in the process
+		// now replicas and engines have been marked for deletion
+
+		if err := vc.ds.WaitEnginesDeleted(volume.Name); err != nil {
```

Comment:
`WaitForVolumeEnginesDeletion` and `WaitForVolumeReplicasDeletion`.

Reply:
Don't do block call here. We should be fine if we just return nil and wait for the next round when replicas or engines haven't been deleted.